CHANGELOG.md

```diff
@@ -17,6 +17,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - OTLP Metrics Exporter now supports the `OTEL_EXPORTER_OTLP_METRICS_TEMPORALITY_PREFERENCE` environment variable. (#4287)
 - Add `WithoutCounterSuffixes` option in `go.opentelemetry.io/otel/exporters/prometheus` to disable addition of `_total` suffixes. (#4306)
 - Add info and debug logging to the metric SDK. (#4315)
+- Handle 204 HTTP status responses in `go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetrichttp` and `go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp`. (#4365)
```

Comment:
Isn't the specification requiring to return `200 OK`?

https://github.com/open-telemetry/opentelemetry-proto/blob/main/docs/specification.md#otlphttp-response

If so then, we should first update the specification.

Reply:
It also mentions that any other HTTP status code should be handled per the HTTP spec.
See my comment in https://github.com/open-telemetry/opentelemetry-go/issues/4363#issuecomment-1651464538

Reply:
I see. The spec mentions it here: https://github.com/open-telemetry/opentelemetry-proto/blob/main/docs/specification.md#all-other-responses

Reply:
The problem is that the specification requires

> the server MUST respond with HTTP 200 OK

While 

> All other HTTP responses that are not explicitly listed in this document should be treated according to HTTP specifications.

is not really specific and it does not seem to be normative.

Have you checked how OTLP exporters in other SDKs (e.g. in Java, .NET, Python) are handling it?

Reply:
Python accepts 200 and 202: https://github.com/open-telemetry/opentelemetry-python/blob/main/exporter/opentelemetry-exporter-otlp-proto-http/src/opentelemetry/exporter/otlp/proto/http/metric_exporter/__init__.py#L191

Java accepts 200 to 299: https://github.com/open-telemetry/opentelemetry-java/blob/main/exporters/common/src/main/java/io/opentelemetry/exporter/internal/http/HttpExporter.java#L85

DotNet relies on `EnsureSuccessStatusCode`, which seems to accept 200 to 299: https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/ExportClient/BaseOtlpHttpExportClient.cs#L58C27-L58C50

Node accepts 200 to 299: https://github.com/open-telemetry/opentelemetry-js/blob/main/experimental/packages/otlp-exporter-base/src/platform/node/util.ts#L102

Reply:
Collector accepts 200 to 299: https://github.com/open-telemetry/opentelemetry-collector/blob/56135230d1d998608cc7c61250b28b64ffdd620b/exporter/otlphttpexporter/otlp.go#L135-L137

Reply:
That's what I was starting to think too.